File: frontend/src/api/matching.ts
import { api } from "@/api/client";

export type MatchSignal = {
  name: string;
  weight: number;
  score: number;
  note: string;
};

export type MatchCandidate = {
  match_id: string;
  contract_id: string;
  project_id: string;
  contract_number: string;
  project_name: string;
  vendor_name: string;
  confidence: number;
  signals: MatchSignal[];
  remaining_budget: string;
  invoiced_to_date: string;
};

export type BoqMappingSuggestion = {
  invoice_line_item_id: string;
  invoice_line_number: number | null;
  boq_item_id: string | null;
  boq_line_number: number | null;
  confidence: number;
  reason: string;
};

export const fetchCandidates = (invoiceId: string): Promise<MatchCandidate[]> =>
  api<MatchCandidate[]>(`/matching/${invoiceId}/candidates`, { method: "POST" });

export const confirmMatch = (invoiceId: string, matchId: string): Promise<unknown> =>
  api(`/matching/${invoiceId}/confirm`, { body: { match_id: matchId } });

export const unlockMatch = (invoiceId: string): Promise<void> =>
  api<void>(`/matching/${invoiceId}/unlock`, { method: "POST" });

export const proposeBoqMapping = (invoiceId: string): Promise<BoqMappingSuggestion[]> =>
  api<BoqMappingSuggestion[]>(`/matching/${invoiceId}/map-boq`, { method: "POST" });

export const applyBoqMapping = (
  invoiceId: string,
  body: BoqMappingSuggestion[],
): Promise<BoqMappingSuggestion[]> =>
  api<BoqMappingSuggestion[]>(`/matching/${invoiceId}/apply-boq-mapping`, { body });

File: backend/app/schemas/matching.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel

from app.schemas.common import ORMModel


class SignalOut(BaseModel):
    name: str
    weight: int
    score: int
    note: str


class MatchCandidateOut(BaseModel):
    match_id: UUID
    contract_id: UUID
    project_id: UUID
    contract_number: str
    project_name: str
    vendor_name: str
    confidence: int
    signals: list[SignalOut]
    remaining_budget: Decimal
    invoiced_to_date: Decimal


class MatchConfirmRequest(BaseModel):
    match_id: UUID


class MatchOut(ORMModel):
    id: UUID
    invoice_id: UUID
    candidate_contract_id: UUID
    confidence: int
    reasoning_json: dict | None
    confirmed: bool
    confirmed_by: UUID | None
    created_at: datetime


class BoqMappingSuggestion(BaseModel):
    invoice_line_item_id: UUID
    invoice_line_number: int | None
    boq_item_id: UUID | None
    boq_line_number: int | None
    confidence: int
    reason: str = ""
